bpo-29546: Improve from-import error message with location

Carreau opened the pull request:
Add location information like canonical module name where identifier
cannot be found and file location if available.

First iteration of this was gh-91

---- 

Testless for now, I can add tests once the exact format has been agreed upon.

Carreau committed:
bpo-29546: Improve from-import error message with location

Add location information like canonical module name where identifier
cannot be found and file location if available.

First iteration of this was gh-91

brettcannon commented on Python/ceval.c:
```diff
@@ -5022,12 +5022,23 @@ import_from(PyObject *v, PyObject *name)
     return x;
  error:
     pkgpath = PyModule_GetFilenameObject(v);
+    if (name == NULL){
```
Missing a space between `)` and `{`.

warsaw reviewed (approved):
Yep, that's exactly the format I was thinking!  Great work.

warsaw commented:
```
Python 3.7.0a0 (default, Feb 15 2017, 14:30:57) 
[GCC 6.3.0 20161229] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from os import floop
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'floop' from 'os' (/home/barry/projects/python/cpython/Lib/os.py)
>>> import snarp
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'snarp'
>>> 
>>> from _opcode import fip
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name 'fip' from '_opcode' (/home/barry/projects/python/cpython/build/lib.linux-x86_64-3.7/_opcode.cpython-37m-x86_64-linux-gnu.so)
>>> 
```

Carreau commented:
> Yep, that's exactly the format I was thinking! Great work.

Good, and thanks.  I'll cleanup and add tests then !

Carreau commented:
Added tests, Misc/NEWS, What's new.

The only thing I'm unsure is what the module's `__file__` extensions is on windows. I'm guessing I can relax the regex in the test. I've assumed `.dll`.

It's also a bit ugly if a module has neither name not a file:

`"cannot import name 'does_not_exist' from '<unknown module name>' (unknown location)"`

But that seem to be rare from how it's currently tested:

```python
class AlwaysAttributeError:
            def __getattr__(self, _):
                raise AttributeError

module_name = 'test_from_import_AttributeError'
self.addCleanup(unload, module_name)
sys.modules[module_name] = AlwaysAttributeError()
```

warsaw commented:
On Feb 15, 2017, at 07:56 PM, Matthias Bussonnier wrote:

>It's also a bit ugly if a module has neither name not a file:
>
>> "cannot import name 'does_not_exist' from '<unknown module name>' (unknown location)"  
>
>But that seem to be rare from how it's currently tested:

Agreed it's ugly, but if it won't happen commonly in practice, it's probably
not worth dealing with specially.  I'm also not sure how you'd print that any
better - it is worth knowing what's unknowable. :)

warsaw commented on Python/ceval.c:
```diff
@@ -5022,12 +5022,23 @@ import_from(PyObject *v, PyObject *name)
+        pkgname_or_unknown = PyUnicode_FromString("<unknown module name>");
+    } else {
+        pkgname_or_unknown = pkgname;
+    }
```
I think there are some reference counting bugs lurking here.

* It's possible the code jumps to `error:` after `pkgname` is decref'd.  I think that decref should be deferred until just before the good-path return.
* `pkgname_or_unknown` will be a new reference if `pkgname` is NULL, or it will steal the `pkgname` reference if it's not.  In either case, the `error` stanza should decref `pkgname_or_unknown`, which will take care of ensuring the object gets decref for any `goto error` path.
* `pkgpath` doesn't get decref'd.  I think you should `Py_XDECREF` it before the `return NULL` just in case `pkgpath` is itself NULL.

Carreau replied:
Thanks for the detailed explanation, I still need to get used to do manual ref counting. 
I've push changes. Let me know if I got things rights.

warsaw replied:
@Carreau Yes, I think the refcounting is right now.  It can sometimes be difficult to trace all the various exit paths to make sure that everything gets decref'd and incref'd properly, but as best I can tell, you've got it right now.  Thanks!

The only other thing is that it's *possible* for the `PyUnicode_FromString("<unknown module name>")` to fail and return NULL.  Probably the only realistic cause of that would be a memory error (not sure if an encoding error could do it given that this is an ascii C string).  If you want to be pedanditc, it would be useful to check for that error condition and do something more reasonable in that case.

Carreau replied:
> If you want to be pedanditc, it would be useful to check for that error condition and do something more reasonable in that case

I'm unsure what "more reasonable" is,  I check and failed with the original error message in that case, but I don't see how if allocating `PyUnicode_FromString("<unknown module name>")` fails we can do something sensible.

warsaw replied:
Probably the best you can do is to just decref any objects that may still have a reference count, and then immediately return NULL.  The PyUnicode_FromString failure would have already set an exception, so that's the one you'll see.  Something like:

```
1 file changed, 4 insertions(+)
Python/ceval.c | 4 ++++

modified   Python/ceval.c
@@ -5024,6 +5024,10 @@ import_from(PyObject *v, PyObject *name)
     pkgpath = PyModule_GetFilenameObject(v);
     if (pkgname == NULL) {
         pkgname_or_unknown = PyUnicode_FromString("<unknown module name>");
+        if (pkgname_or_unknown == NULL) {
+            Py_XDECREF(pkgpath);
+            return NULL;
+        }
     } else {
         pkgname_or_unknown = pkgname;
     }
```

warsaw reviewed (changes requested):
Please review the potential reference counting bugs.

Carreau committed:
Update reference counting

from Barry Warsaw:

- It's possible the code jumps to error: after pkgname is decref'd. I
  think that decref should be deferred until just before the good-path
  return.

- pkgname_or_unknown will be a new reference if pkgname is NULL, or it
  will steal the pkgname reference if it's not. In either case, the error
  stanza should decref pkgname_or_unknown, which will take care of
  ensuring the object gets decref for any goto error path.

- pkgpath doesn't get decref'd. I think you should Py_XDECREF it before
  the return NULL just in case pkgpath is itself NULL.

warsaw commented on Python/ceval.c:
```diff
@@ -5018,18 +5017,32 @@ import_from(PyObject *v, PyObject *name)
     }
 
+    Py_XDECREF(pkgname_or_unknown)
+    Py_XDECREF(pkgpath)
```
Oh whoops, these lines need to end with semicolons.

warsaw reviewed (changes requested):
Semicolons to end lines.

Please check for `PyUnicode_FromString()` returning NULL.

Carreau committed:
Handle if PyUnicode_FromString fails.

Carreau commented:
semicolon added

warsaw commented on Python/ceval.c:
```diff
@@ -5018,18 +5017,35 @@ import_from(PyObject *v, PyObject *name)
+    }
 
-    if (pkgpath == NULL || !PyUnicode_Check(pkgpath)) {
+    if (pkgname_or_unknown == NULL){
```
You need a space between the paren and open brace.

warsaw reviewed (changes requested):
Two more small comments, but I think this is really close!

Carreau commented:
Comments should be addressed. Thanks !

warsaw commented:
@Carreau Thanks, this LGTM now.  I'm pushing the big green button.

Carreau commented:
Thanks !
